include/ddsrouter/user_interface/impl/SignalHandler.ipp

```diff
@@ -0,0 +1,153 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file SignalHandler.ipp
+ *
+ */
+
+#ifndef _DDSROUTER_USERINTERFACE_IMPL_SIGNALHANDLER_IPP_
+#define _DDSROUTER_USERINTERFACE_IMPL_SIGNALHANDLER_IPP_
+
+#include <algorithm>
+#include <csignal>
+
+#include <ddsrouter/types/Log.hpp>
+#include <ddsrouter/user_interface/SignalHandler.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+namespace ui {
+
+template <int SigNum>
+std::vector<SignalHandler<SigNum>*> SignalHandler<SigNum>::active_handlers_;
+
+template <int SigNum>
+std::mutex SignalHandler<SigNum>::active_handlers_mutex_;
+
+template <int SigNum>
+SignalHandler<SigNum>::SignalHandler() noexcept
+    : SignalHandler<SigNum>(
+        [](int signum)
+        {
+            logInfo(DDSROUTER_SIGNALHANDLER,
+                "Received signal " << signum << " in specific handler .");
+        })
+{
+}
+
+template <int SigNum>
+SignalHandler<SigNum>::SignalHandler(std::function<void(int)> callback) noexcept
+    : EventHandler<int>(callback)
+{
+    add_to_active_handlers_();
+}
+
+template <int SigNum>
+SignalHandler<SigNum>::~SignalHandler()
+{
+    erase_from_active_handlers_();
+}
+
+template <int SigNum>
+void SignalHandler<SigNum>::callback_set_() noexcept
+{
+    add_to_active_handlers_();
+}
+
+template <int SigNum>
+void SignalHandler<SigNum>::callback_unset_() noexcept
+{
+    erase_from_active_handlers_();
+}
+
+template <int SigNum>
+void SignalHandler<SigNum>::add_to_active_handlers_() noexcept
+{
+    std::lock_guard<std::mutex> lock(active_handlers_mutex_);
+
+    logInfo(DDSROUTER_SIGNALHANDLER,
+        "Add signal handler to signal " << SigNum << ".");
+
+    active_handlers_.push_back(this);
+
+    // First value included, set signal handler
+    if (active_handlers_.size() == 1)
+    {
+        set_signal_handler_();
+    }
+}
+
+template <int SigNum>
+void SignalHandler<SigNum>::erase_from_active_handlers_() noexcept
+{
+    std::lock_guard<std::mutex> lock(active_handlers_mutex_);
+
+    logInfo(DDSROUTER_SIGNALHANDLER,
+        "Erase signal handler from signal " << SigNum << ".");
+
+    active_handlers_.erase(
+        std::remove(
+            active_handlers_.begin(),
+            active_handlers_.end(),
+            this),
+        active_handlers_.end());
+
+    // Last handler erased, unset signal handler
+    if (active_handlers_.size() == 0)
+    {
+        unset_signal_handler_();
+    }
+}
+
+template <int SigNum>
+void SignalHandler<SigNum>::signal_handler_routine_(int signum) noexcept
+{
+    std::lock_guard<std::mutex> lock(active_handlers_mutex_);
+
+    logInfo(DDSROUTER_SIGNALHANDLER,
+        "Received signal " << signum << ".");
+
+    if (signum != SigNum)
+    {
+        logError(DDSROUTER_SIGNALHANDLER, "Signal handler associated to incorrect signal,");
```

Comment:
Even if it should never happen, maybe the function should end here.

Reply:
:+1: